File: ui/NotificationsScreen.kt
package com.example.healthride.ui

import androidx.compose.animation.*
import androidx.compose.animation.core.*
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material.icons.outlined.*
import androidx.compose.material.icons.rounded.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.*
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.*
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.example.healthride.ui.theme.*
import kotlinx.coroutines.delay
import java.text.SimpleDateFormat
import java.util.*

data class Notification(
    val id: String,
    val title: String,
    val message: String,
    val time: Date,
    val type: NotificationType,
    val isRead: Boolean
)

enum class NotificationType {
    RIDE_CONFIRMED, DRIVER_ASSIGNED, RIDE_REMINDER, RIDE_COMPLETED, PROMOTION
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun NotificationsScreen(
    onBackClick: () -> Unit = {}
) {
    // Sample notifications
    val notifications = remember {
        listOf(
            Notification(
                id = "n1",
                title = "Driver Assigned",
                message = "Michael Johnson will be your driver for your upcoming ride to Medical Center.",
                time = Calendar.getInstance().apply { add(Calendar.HOUR, -1) }.time,
                type = NotificationType.DRIVER_ASSIGNED,
                isRead = false
            ),
            Notification(
                id = "n2",
                title = "Ride Reminder",
                message = "Reminder: You have a scheduled ride tomorrow at 9:30 AM to City Hospital.",
                time = Calendar.getInstance().apply { add(Calendar.HOUR, -5) }.time,
                type = NotificationType.RIDE_REMINDER,
                isRead = false
            ),
            Notification(
                id = "n3",
                title = "Ride Completed",
                message = "Your ride to Medical Center has been completed. Thank you for using HealthRide!",
                time = Calendar.getInstance().apply { add(Calendar.DAY_OF_MONTH, -1) }.time,
                type = NotificationType.RIDE_COMPLETED,
                isRead = true
            ),
            Notification(
                id = "n4",
                title = "Special Offer",
                message = "Get 20% off on your next 3 rides! Offer valid for the next 7 days.",
                time = Calendar.getInstance().apply { add(Calendar.DAY_OF_MONTH, -2) }.time,
                type = NotificationType.PROMOTION,
                isRead = true
            ),
            Notification(
                id = "n5",
                title = "Ride Confirmed",
                message = "Your ride to Wellness Clinic has been confirmed for next Monday.",
                time = Calendar.getInstance().apply { add(Calendar.DAY_OF_MONTH, -3) }.time,
                type = NotificationType.RIDE_CONFIRMED,
                isRead = true
            )
        )
    }

    // Animation for content appearance
    var showContent by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) {
        delay(100)
        showContent = true
    }

    Box(modifier = Modifier.fillMaxSize()) {
        // Beautiful gradient background
        Canvas(modifier = Modifier.fillMaxSize()) {
            val gradientBrush = Brush.linearGradient(
                colors = listOf(
                    Color(0xFFF0F4FF),  // Light blue
                    Color(0xFFF5F0FF),  // Light purple
                    Color(0xFFF8FBFF)   // White-ish
                ),
                start = Offset(0f, 0f),
                end = Offset(size.width, size.height)
            )
            drawRect(brush = gradientBrush)

            // Abstract decorative shape
            val accentBrush = Brush.radialGradient(
                colors = listOf(
                    PrimaryBlue.copy(alpha = 0.1f),
                    Color.Transparent
                ),
                center = Offset(size.width * 0.1f, size.height * 0.1f),
                radius = size.width * 0.5f
            )
            drawCircle(
                brush = accentBrush,
                radius = size.width * 0.5f,
                center = Offset(size.width * 0.1f, size.height * 0.1f)
            )

            val accentBrush2 = Brush.radialGradient(
                colors = listOf(
                    PrimaryPurple.copy(alpha = 0.08f),
                    Color.Transparent
                ),
                center = Offset(size.width * 0.9f, size.height * 0.7f),
                radius = size.width * 0.6f
            )
            drawCircle(
                brush = accentBrush2,
                radius = size.width * 0.6f,
                center = Offset(size.width * 0.9f, size.height * 0.7f)
            )
        }

        Scaffold(
            topBar = {
                TopAppBar(
                    title = { Text("Notifications", fontWeight = FontWeight.SemiBold) },
                    navigationIcon = {
                        IconButton(onClick = onBackClick) {
                            Icon(
                                imageVector = Icons.Default.ArrowBack,
                                contentDescription = "Back"
                            )
                        }
                    },
                    actions = {
                        // Mark all as read button with animation
                        IconButton(
                            onClick = { /* Mark all as read */ }
                        ) {
                            Icon(
                                imageVector = Icons.Outlined.DoneAll,
                                contentDescription = "Mark all as read",
                                tint = PrimaryBlue
                            )
                        }
                    },
                    colors = TopAppBarDefaults.topAppBarColors(
                        containerColor = Color.Transparent,
                        titleContentColor = TextDarkBlue
                    )
                )
            },
            containerColor = Color.Transparent
        ) { innerPadding ->
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(top = innerPadding.calculateTopPadding())
            ) {
                // Animated content appearance
                AnimatedVisibility(
                    visible = showContent,
                    enter = fadeIn() + slideInVertically(initialOffsetY = { it / 3 })
                ) {
                    if (notifications.isEmpty()) {
                        // Empty state with animation
                        ColorfulEmptyState()
                    } else {
                        Column(modifier = Modifier.fillMaxSize()) {
                            // Unread count badge
                            val unreadCount = notifications.count { !it.isRead }
                            if (unreadCount > 0) {
                                GradientUnreadBanner(unreadCount = unreadCount)
                            }

                            // Group notifications by day
                            val todayStart = Calendar.getInstance().apply {
                                set(Calendar.HOUR_OF_DAY, 0)
                                set(Calendar.MINUTE, 0)
                                set(Calendar.SECOND, 0)
                                set(Calendar.MILLISECOND, 0)
                            }.time

                            val yesterdayStart = Calendar.getInstance().apply {
                                add(Calendar.DAY_OF_YEAR, -1)
                                set(Calendar.HOUR_OF_DAY, 0)
                                set(Calendar.MINUTE, 0)
                                set(Calendar.SECOND, 0)
                                set(Calendar.MILLISECOND, 0)
                            }.time

                            val groupedNotifications = notifications.groupBy { notification ->
                                when {
                                    notification.time.after(todayStart) -> "Today"
                                    notification.time.after(yesterdayStart) -> "Yesterday"
                                    else -> "Earlier"
                                }
                            }

                            LazyColumn(
                                modifier = Modifier
                                    .fillMaxSize()
                                    .padding(horizontal = 16.dp),
                                contentPadding = PaddingValues(bottom = 16.dp)
                            ) {
                                groupedNotifications.forEach { (dayGroup, notificationsInGroup) ->
                                    // Date header
                                    item {
                                        Row(
                                            verticalAlignment = Alignment.CenterVertically,
                                            modifier = Modifier
                                                .fillMaxWidth()
                                                .padding(vertical = 16.dp, horizontal = 4.dp)
                                        ) {
                                            Text(
                                                text = dayGroup,
                                                style = MaterialTheme.typography.titleMedium,
                                                fontWeight = FontWeight.Bold,
                                                color = TextDarkBlue
                                            )

                                            Spacer(modifier = Modifier.width(12.dp))

                                            Divider(
                                                modifier = Modifier.weight(1f),
                                                color = TextLightGray.copy(alpha = 0.5f)
                                            )
                                        }
                                    }

                                    // Notifications for this day
                                    items(notificationsInGroup) { notification ->
                                        // Staggered animation
                                        var itemVisible by remember { mutableStateOf(false) }
                                        LaunchedEffect(notification.id) {
                                            delay(100)
                                            itemVisible = true
                                        }

                                        AnimatedVisibility(
                                            visible = itemVisible,
                                            enter = fadeIn() + expandVertically(
                                                expandFrom = Alignment.Top,
                                                animationSpec = spring(
                                                    dampingRatio = Spring.DampingRatioMediumBouncy,
                                                    stiffness = Spring.StiffnessLow
                                                )
                                            )
                                        ) {
                                            ColorfulNotificationCard(notification = notification)
                                        }

                                        Spacer(modifier = Modifier.height(12.dp))
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

@Composable
fun GradientUnreadBanner(unreadCount: Int) {
    Card(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 8.dp)
            .shadow(
                elevation = 8.dp,
                shape = RoundedCornerShape(16.dp),
                spotColor = PrimaryBlue.copy(alpha = 0.2f)
            ),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(containerColor = Color.Transparent)
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .background(
                    brush = Brush.linearGradient(
                        colors = listOf(
                            PrimaryBlue.copy(alpha = 0.9f),
                            PrimaryPurple.copy(alpha = 0.9f)
                        )
                    )
                )
                .padding(horizontal = 16.dp, vertical = 14.dp)
        ) {
            Row(
                verticalAlignment = Alignment.CenterVertically
            ) {
                // Pulsating counter
                val infiniteTransition = rememberInfiniteTransition(label = "pulse_animation")
                val pulsate = infiniteTransition.animateFloat(
                    initialValue = 0.9f,
                    targetValue = 1.1f,
                    animationSpec = infiniteRepeatable(
                        animation = tween(1000, easing = EaseInOut),
                        repeatMode = RepeatMode.Reverse
                    ),
                    label = "pulse"
                )

                Box(
                    modifier = Modifier
                        .size(32.dp)
                        .graphicsLayer {
                            scaleX = pulsate.value
                            scaleY = pulsate.value
                        }
                        .clip(CircleShape)
                        .background(Color.White),
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        text = unreadCount.toString(),
                        color = PrimaryBlue,
                        fontWeight = FontWeight.Bold
                    )
                }

                Spacer(modifier = Modifier.width(16.dp))

                Text(
                    text = "You have $unreadCount unread notification${if (unreadCount > 1) "s" else ""}",
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.Medium,
                    color = Color.White
                )
            }
        }
    }
}

@Composable
fun ColorfulNotificationCard(notification: Notification) {
    val (iconColor, gradientStart, gradientEnd) = when (notification.type) {
        NotificationType.RIDE_CONFIRMED -> Triple(
            SuccessGreen,
            SuccessGreen,
            Color(0xFF34D399)
        )
        NotificationType.DRIVER_ASSIGNED -> Triple(
            PrimaryBlue,
            PrimaryBlue,
            Color(0xFF4CC2FF)
        )
        NotificationType.RIDE_REMINDER -> Triple(
            WarningYellow,
            WarningYellow,
            Color(0xFFFCD34D)
        )
        NotificationType.RIDE_COMPLETED -> Triple(
            SuccessGreen,
            SuccessGreen,
            Color(0xFF34D399)
        )
        NotificationType.PROMOTION -> Triple(
            PrimaryPurple,
            PrimaryPurple,
            Color(0xFF9F6EFF)
        )
    }

    val icon = when (notification.type) {
        NotificationType.RIDE_CONFIRMED -> Icons.Rounded.CheckCircle
        NotificationType.DRIVER_ASSIGNED -> Icons.Rounded.Person
        NotificationType.RIDE_REMINDER -> Icons.Rounded.Alarm
        NotificationType.RIDE_COMPLETED -> Icons.Rounded.DoneAll
        NotificationType.PROMOTION -> Icons.Rounded.LocalOffer
    }

    val timeFormat = SimpleDateFormat("h:mm a", Locale.getDefault())

    Card(
        modifier = Modifier
            .fillMaxWidth()
            .shadow(
                elevation = if (notification.isRead) 4.dp else 8.dp,
                shape = RoundedCornerShape(20.dp),
                spotColor = if (notification.isRead)
                    Color.Gray.copy(alpha = 0.1f)
                else
                    gradientStart.copy(alpha = 0.2f)
            )
            .graphicsLayer {
                alpha = if (notification.isRead) 0.85f else 1f
            },
        shape = RoundedCornerShape(20.dp),
        colors = CardDefaults.cardColors(containerColor = Color.White)
    ) {
        Row(
            modifier = Modifier.padding(16.dp),
            verticalAlignment = Alignment.Top
        ) {
            // Colorful icon with gradient background
            Box(
                modifier = Modifier
                    .size(50.dp)
                    .clip(CircleShape)
                    .background(
                        brush = Brush.linearGradient(
                            colors = listOf(gradientStart, gradientEnd)
                        )
                    ),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = icon,
                    contentDescription = null,
                    tint = Color.White,
                    modifier = Modifier.size(24.dp)
                )
            }

            Spacer(modifier = Modifier.width(16.dp))

            Column(modifier = Modifier.weight(1f)) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = notification.title,
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = if (notification.isRead) FontWeight.Medium else FontWeight.Bold,
                        color = TextDarkBlue
                    )

                    // Unread indicator
                    if (!notification.isRead) {
                        Box(
                            modifier = Modifier
                                .size(8.dp)
                                .clip(CircleShape)
                                .background(
                                    brush = Brush.linearGradient(
                                        colors = listOf(gradientStart, gradientEnd)
                                    )
                                )
                        )
                    }
                }

                Spacer(modifier = Modifier.height(4.dp))

                Text(
                    text = notification.message,
                    style = MaterialTheme.typography.bodyMedium,
                    color = if (notification.isRead) TextMediumGray else TextDarkBlue,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis
                )

                Spacer(modifier = Modifier.height(8.dp))

                Box(
                    modifier = Modifier
                        .align(Alignment.End)
                        .clip(RoundedCornerShape(50.dp))
                        .background(Color(0xFFF0F4FF))
                        .padding(horizontal = 10.dp, vertical = 4.dp)
                ) {
                    Text(
                        text = timeFormat.format(notification.time),
                        style = MaterialTheme.typography.labelSmall,
                        color = TextMediumGray
                    )
                }
            }
        }
    }
}

@Composable
fun ColorfulEmptyState() {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        // Animated empty state
        val infiniteTransition = rememberInfiniteTransition(label = "floating_animation")
        val translateY = infiniteTransition.animateFloat(
            initialValue = 0f,
            targetValue = 20f,
            animationSpec = infiniteRepeatable(
                animation = tween(2000, easing = EaseInOut),
                repeatMode = RepeatMode.Reverse
            ),
            label = "float"
        )

        Box(
            modifier = Modifier
                .size(160.dp)
                .graphicsLayer {
                    translationY = translateY.value
                }
                .shadow(
                    elevation = 16.dp,
                    shape = CircleShape,
                    spotColor = PrimaryBlue.copy(alpha = 0.3f)
                )
                .clip(CircleShape)
                .background(
                    brush = Brush.linearGradient(
                        colors = listOf(
                            PrimaryBlue.copy(alpha = 0.7f),
                            PrimaryPurple.copy(alpha = 0.7f)
                        )
                    )
                ),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = Icons.Outlined.Notifications,
                contentDescription = null,
                tint = Color.White,
                modifier = Modifier.size(80.dp)
            )
        }

        Spacer(modifier = Modifier.height(32.dp))

        Text(
            text = "No Notifications Yet",
            style = MaterialTheme.typography.headlineMedium,
            fontWeight = FontWeight.Bold,
            color = TextDarkBlue
        )

        Spacer(modifier = Modifier.height(16.dp))

        Text(
            text = "We'll notify you when you have new messages, ride updates, or special offers",
            style = MaterialTheme.typography.bodyLarge,
            color = TextMediumGray,
            textAlign = androidx.compose.ui.text.style.TextAlign.Center
        )

        Spacer(modifier = Modifier.height(32.dp))

        // Beautiful gradient button
        Card(
            modifier = Modifier
                .shadow(
                    elevation = 8.dp,
                    shape = RoundedCornerShape(50.dp),
                    spotColor = PrimaryBlue.copy(alpha = 0.3f)
                ),
            shape = RoundedCornerShape(50.dp),
            colors = CardDefaults.cardColors(containerColor = Color.Transparent)
        ) {
            Box(
                modifier = Modifier
                    .background(
                        brush = Brush.horizontalGradient(
                            colors = listOf(
                                PrimaryBlue,
                                PrimaryPurple
                            )
                        )
                    )
                    .padding(horizontal = 32.dp, vertical = 16.dp)
            ) {
                Text(
                    text = "Book a Ride",
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                    color = Color.White
                )
            }
        }
    }
}

File: lib/src/screens/notifications_screen.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import '../constants/app_color.dart';
import '../widgets/components/gradient_button.dart';
import 'dart:math' as math;

/// Notification model
class NotificationItem {
  /// Unique identifier
  final String id;

  /// Notification title
  final String title;

  /// Notification message content
  final String message;

  /// Time the notification was received
  final DateTime time;

  /// Type of notification
  final NotificationType type;

  /// Whether the notification has been read
  final bool isRead;

  /// Creates a notification
  const NotificationItem({
    required this.id,
    required this.title,
    required this.message,
    required this.time,
    required this.type,
    required this.isRead,
  });
}

/// Types of notifications
enum NotificationType {
  /// Ride booking confirmed
  rideConfirmed,

  /// Driver has been assigned to the ride
  driverAssigned,

  /// Reminder about upcoming ride
  rideReminder,

  /// Ride has been completed
  rideCompleted,

  /// Special promotion or offer
  promotion,
}

/// Screen that displays user notifications
class NotificationsScreen extends StatefulWidget {
  /// Creates a notifications screen
  const NotificationsScreen({super.key});

  @override
  State<NotificationsScreen> createState() => _NotificationsScreenState();
}

class _NotificationsScreenState extends State<NotificationsScreen>
    with SingleTickerProviderStateMixin {
  bool _showContent = false;
  late List<NotificationItem> notifications;

  @override
  void initState() {
    super.initState();
    _loadNotifications();

    // Start animations
    Future.delayed(const Duration(milliseconds: 100), () {
      if (mounted) {
        setState(() {
          _showContent = true;
        });
      }
    });
  }

  void _loadNotifications() {
    // Sample notifications data
    notifications = [
      NotificationItem(
        id: "n1",
        title: "Driver Assigned",
        message:
            "Michael Johnson will be your driver for your upcoming ride to Medical Center.",
        time: DateTime.now().subtract(const Duration(hours: 1)),
        type: NotificationType.driverAssigned,
        isRead: false,
      ),
      NotificationItem(
        id: "n2",
        title: "Ride Reminder",
        message:
            "Reminder: You have a scheduled ride tomorrow at 9:30 AM to City Hospital.",
        time: DateTime.now().subtract(const Duration(hours: 5)),
        type: NotificationType.rideReminder,
        isRead: false,
      ),
      NotificationItem(
        id: "n3",
        title: "Ride Completed",
        message:
            "Your ride to Medical Center has been completed. Thank you for using HealthRide!",
        time: DateTime.now().subtract(const Duration(days: 1)),
        type: NotificationType.rideCompleted,
        isRead: true,
      ),
      NotificationItem(
        id: "n4",
        title: "Special Offer",
        message:
            "Get 20% off on your next 3 rides! Offer valid for the next 7 days.",
        time: DateTime.now().subtract(const Duration(days: 2)),
        type: NotificationType.promotion,
        isRead: true,
      ),
      NotificationItem(
        id: "n5",
        title: "Ride Confirmed",
        message:
            "Your ride to Wellness Clinic has been confirmed for next Monday.",
        time: DateTime.now().subtract(const Duration(days: 3)),
        type: NotificationType.rideConfirmed,
        isRead: true,
      ),
    ];
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      extendBodyBehindAppBar: true,
      appBar: AppBar(
        title: const Text(
          'Notifications',
          style: TextStyle(
            fontWeight: FontWeight.w600,
            color: AppColor.textDarkBlue,
          ),
        ),
        backgroundColor: Colors.transparent,
        elevation: 0,
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          color: AppColor.textDarkBlue,
          onPressed: () {
            Navigator.pop(context);
          },
        ),
        actions: [
          IconButton(
            icon: const Icon(Icons.done_all_outlined),
            color: AppColor.primaryBlue,
            onPressed: () {
              // Mark all as read
            },
          ),
        ],
      ),
      body: Container(
        decoration: const BoxDecoration(
          gradient: LinearGradient(
            begin: Alignment.topLeft,
            end: Alignment.bottomRight,
            colors: [
              Color(0xFFF0F4FF), // Light blue
              Color(0xFFF5F0FF), // Light purple
              Color(0xFFF8FBFF), // White-ish
            ],
          ),
        ),
        child: Stack(
          children: [
            // Decorative shape top-left
            Positioned(
              top: 0,
              left: 0,
              child: Container(
                width: MediaQuery.of(context).size.width * 0.8,
                height: MediaQuery.of(context).size.width * 0.8,
                decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  gradient: RadialGradient(
                    colors: [
                      AppColor.primaryBlue.withOpacity(0.1),
                      Colors.transparent,
                    ],
                  ),
                ),
              ),
            ),

            // Decorative shape bottom-right
            Positioned(
              bottom: 0,
              right: 0,
              child: Container(
                width: MediaQuery.of(context).size.width * 0.9,
                height: MediaQuery.of(context).size.width * 0.9,
                decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  gradient: RadialGradient(
                    colors: [
                      AppColor.primaryPurple.withOpacity(0.08),
                      Colors.transparent,
                    ],
                  ),
                ),
              ),
            ),

            SafeArea(
              child: AnimatedOpacity(
                opacity: _showContent ? 1.0 : 0.0,
                duration: const Duration(milliseconds: 500),
                curve: Curves.easeOut,
                child: AnimatedSlide(
                  offset: _showContent ? Offset.zero : const Offset(0, 0.1),
                  duration: const Duration(milliseconds: 500),
                  curve: Curves.easeOutCubic,
                  child: notifications.isEmpty
                      ? _buildEmptyState()
                      : _buildNotificationsList(),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildNotificationsList() {
    final unreadCount = notifications.where((n) => !n.isRead).length;

    // Group notifications by day
    final todayDate = DateTime(
      DateTime.now().year,
      DateTime.now().month,
      DateTime.now().day,
    );
    final yesterdayDate = todayDate.subtract(const Duration(days: 1));

    final groupedNotifications = <String, List<NotificationItem>>{};

    for (final notification in notifications) {
      final notificationDate = DateTime(
        notification.time.year,
        notification.time.month,
        notification.time.day,
      );

      String group;
      if (notificationDate == todayDate) {
        group = 'Today';
      } else if (notificationDate == yesterdayDate) {
        group = 'Yesterday';
      } else {
        group = 'Earlier';
      }

      if (!groupedNotifications.containsKey(group)) {
        groupedNotifications[group] = [];
      }
      groupedNotifications[group]!.add(notification);
    }

    return Column(
      children: [
        // Unread count banner
        if (unreadCount > 0) _buildUnreadBanner(unreadCount),

        // Notifications list
        Expanded(
          child: ListView.builder(
            padding: const EdgeInsets.all(16),
            itemCount: groupedNotifications.keys.length,
            itemBuilder: (context, index) {
              final dateGroup = groupedNotifications.keys.elementAt(index);
              final notificationsForGroup = groupedNotifications[dateGroup]!;

              return Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  // Date header
                  Row(
                    children: [
                      Text(
                        dateGroup,
                        style: const TextStyle(
                          fontSize: 16,
                          fontWeight: FontWeight.bold,
                          color: AppColor.textDarkBlue,
                        ),
                      ),
                      const SizedBox(width: 12),
                      Expanded(
                        child: Divider(
                          color: AppColor.textLightGray.withOpacity(0.5),
                        ),
                      ),
                    ],
                  ),
                  const SizedBox(height: 16),

                  // Notifications for this group
                  ...notificationsForGroup
                      .map((notification) => Padding(
                            padding: const EdgeInsets.only(bottom: 12),
                            child: TweenAnimationBuilder<double>(
                              tween: Tween(begin: 0.0, end: 1.0),
                              duration: Duration(
                                  milliseconds: 300 +
                                      (notificationsForGroup
                                              .indexOf(notification) *
                                          100)),
                              builder: (context, value, child) {
                                return Opacity(
                                  opacity: value,
                                  child: Transform.translate(
                                    offset: Offset(0, 20 * (1 - value)),
                                    child: child,
                                  ),
                                );
                              },
                              child: _buildNotificationCard(notification),
                            ),
                          ))
                      .toList(),

                  const SizedBox(height: 16),
                ],
              );
            },
          ),
        ),
      ],
    );
  }

  Widget _buildUnreadBanner(int unreadCount) {
    return Container(
      margin: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(16),
        gradient: const LinearGradient(
          colors: [
            AppColor.primaryBlue,
            AppColor.primaryPurple,
          ],
        ),
        boxShadow: [
          BoxShadow(
            color: AppColor.primaryBlue.withOpacity(0.3),
            blurRadius: 8,
            offset: const Offset(0, 4),
          ),
        ],
      ),
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 14),
      child: Row(
        children: [
          // Pulsating counter with animation
          TweenAnimationBuilder<double>(
            tween: Tween(begin: 0.9, end: 1.1),
            duration: const Duration(milliseconds: 1000),
            curve: Curves.easeInOut,
            builder: (context, value, child) {
              return Transform.scale(
                scale: value,
                child: child,
              );
            },
            child: Container(
              width: 32,
              height: 32,
              decoration: const BoxDecoration(
                color: Colors.white,
                shape: BoxShape.circle,
              ),
              alignment: Alignment.center,
              child: Text(
                unreadCount.toString(),
                style: const TextStyle(
                  color: AppColor.primaryBlue,
                  fontWeight: FontWeight.bold,
                ),
              ),
            ),
          ),
          const SizedBox(width: 16),
          Text(
            "You have $unreadCount unread notification${unreadCount > 1 ? 's' : ''}",
            style: const TextStyle(
              color: Colors.white,
              fontWeight: FontWeight.w500,
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildNotificationCard(NotificationItem notification) {
    // Define colors based on notification type
    final Color primaryColor;
    final Color secondaryColor;
    final IconData icon;

    switch (notification.type) {
      case NotificationType.rideConfirmed:
        primaryColor = AppColor.successGreen;
        secondaryColor = const Color(0xFF34D399);
        icon = Icons.check_circle_rounded;
        break;
      case NotificationType.driverAssigned:
        primaryColor = AppColor.primaryBlue;
        secondaryColor = const Color(0xFF4CC2FF);
        icon = Icons.person_rounded;
        break;
      case NotificationType.rideReminder:
        primaryColor = AppColor.warningYellow;
        secondaryColor = const Color(0xFFFCD34D);
        icon = Icons.alarm_rounded;
        break;
      case NotificationType.rideCompleted:
        primaryColor = AppColor.successGreen;
        secondaryColor = const Color(0xFF34D399);
        icon = Icons.done_all_rounded;
        break;
      case NotificationType.promotion:
        primaryColor = AppColor.primaryPurple;
        secondaryColor = const Color(0xFF9F6EFF);
        icon = Icons.local_offer_rounded;
        break;
    }

    // Format time
    final timeFormat = DateFormat('h:mm a');
    final formattedTime = timeFormat.format(notification.time);

    return Container(
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(20),
        boxShadow: [
          BoxShadow(
            color: notification.isRead
                ? Colors.grey.withOpacity(0.1)
                : primaryColor.withOpacity(0.2),
            blurRadius: notification.isRead ? 4 : 8,
            offset: const Offset(0, 2),
          ),
        ],
      ),
      child: Opacity(
        opacity: notification.isRead ? 0.85 : 1.0,
        child: Padding(
          padding: const EdgeInsets.all(16),
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              // Icon with gradient background
              Container(
                width: 50,
                height: 50,
                decoration: BoxDecoration(
                  shape: BoxShape.circle,
                  gradient: LinearGradient(
                    colors: [primaryColor, secondaryColor],
                  ),
                ),
                child: Icon(
                  icon,
                  color: Colors.white,
                  size: 24,
                ),
              ),
              const SizedBox(width: 16),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: [
                        Text(
                          notification.title,
                          style: TextStyle(
                            fontSize: 16,
                            fontWeight: notification.isRead
                                ? FontWeight.w500
                                : FontWeight.bold,
                            color: AppColor.textDarkBlue,
                          ),
                        ),
                        if (!notification.isRead)
                          Container(
                            width: 8,
                            height: 8,
                            decoration: BoxDecoration(
                              shape: BoxShape.circle,
                              gradient: LinearGradient(
                                colors: [primaryColor, secondaryColor],
                              ),
                            ),
                          ),
                      ],
                    ),
                    const SizedBox(height: 4),
                    Text(
                      notification.message,
                      style: TextStyle(
                        fontSize: 14,
                        color: notification.isRead
                            ? AppColor.textMediumGray
                            : AppColor.textDarkBlue,
                      ),
                      maxLines: 2,
                      overflow: TextOverflow.ellipsis,
                    ),
                    const SizedBox(height: 8),
                    Align(
                      alignment: Alignment.centerRight,
                      child: Container(
                        padding: const EdgeInsets.symmetric(
                          horizontal: 10,
                          vertical: 4,
                        ),
                        decoration: BoxDecoration(
                          color: const Color(0xFFF0F4FF),
                          borderRadius: BorderRadius.circular(50),
                        ),
                        child: Text(
                          formattedTime,
                          style: const TextStyle(
                            fontSize: 12,
                            color: AppColor.textMediumGray,
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildEmptyState() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          // Animated floating icon
          TweenAnimationBuilder<double>(
            tween: Tween(begin: 0, end: 1),
            duration: const Duration(milliseconds: 1000),
            builder: (context, value, child) {
              return Transform.translate(
                offset: Offset(0, math.sin(value * math.pi * 2) * 10),
                child: Container(
                  width: 160,
                  height: 160,
                  decoration: BoxDecoration(
                    shape: BoxShape.circle,
                    gradient: const LinearGradient(
                      colors: [
                        AppColor.primaryBlue,
                        AppColor.primaryPurple,
                      ],
                    ),
                    boxShadow: [
                      BoxShadow(
                        color: AppColor.primaryBlue.withOpacity(0.3),
                        blurRadius: 16,
                        offset: const Offset(0, 8),
                      ),
                    ],
                  ),
                  child: const Icon(
                    Icons.notifications_outlined,
                    size: 80,
                    color: Colors.white,
                  ),
                ),
              );
            },
          ),
          const SizedBox(height: 32),
          const Text(
            "No Notifications Yet",
            style: TextStyle(
              fontSize: 24,
              fontWeight: FontWeight.bold,
              color: AppColor.textDarkBlue,
            ),
          ),
          const SizedBox(height: 16),
          const Padding(
            padding: EdgeInsets.symmetric(horizontal: 32),
            child: Text(
              "We'll notify you when you have new messages, ride updates, or special offers",
              textAlign: TextAlign.center,
              style: TextStyle(
                fontSize: 16,
                color: AppColor.textMediumGray,
              ),
            ),
          ),
          const SizedBox(height: 32),
          GradientButton(
            text: 'Book a Ride',
            onPressed: () {
              // Handle booking
            },
            width: 200,
          ),
        ],
      ),
    );
  }
}
